Bugfix: Add Content Encoding to write attachment method call

swcarter007 opened the pull request:
### Context

Problem described in Issues
[Image not saving correctly](https://github.com/allure-framework/allure-js/issues/467)


#### Checklist
- [X] [Sign Allure CLA][cla]
- [X] Provide unit tests

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/allure-framework/allure-js?pullRequest=468) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/allure-framework/allure-js?pullRequest=468) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/allure-framework/allure-js?pullRequest=468) it.</sub>

swcarter007 commented on packages/allure-cucumberjs/src/BufferEncodingUtility.ts:
```diff
@@ -0,0 +1,8 @@
+    case "BASE64":
+      return "base64";
+    default:
+    return "utf-8";
```
I'm not entirely sure what the base buffer encoding should be if its not set.  CucumberJs only had two encoding options, "BASE64" and "IDENTITY".  The default node encoding is "utf-8" I believe, so I set the default here if it wasn't previously set.

swcarter007 replied:
I removed this function as Buffer.isEncoding() does the check for a valid type automatically.

swcarter007 commented on packages/allure-cucumberjs/test/specs/allure_cucumberjs_test.ts:
```diff
@@ -81,12 +81,22 @@ const dataSet: { [name: string]: ITestFormatterOptions } = {
       });
+      Given("add an image", function () {
+        // example base64 encoded image for testing
+        const base64Image = "iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAApgAAAKYB3X3/OAAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAANCSURBVEiJtZZPbBtFFMZ/M7ubXdtdb1xSFyeilBapySVU8h8OoFaooFSqiihIVIpQBKci6KEg9Q6H9kovIHoCIVQJJCKE1ENFjnAgcaSGC6rEnxBwA04Tx43t2FnvDAfjkNibxgHxnWb2e/u992bee7tCa00YFsffekFY+nUzFtjW0LrvjRXrCDIAaPLlW0nHL0SsZtVoaF98mLrx3pdhOqLtYPHCha ...
```
I used the codepen image taken from this example.
https://codepen.io/jamiekane/pen/YayWOa

baev replied:
please confirm that this image is licensed under Apache 2 compatible license

swcarter007 replied:
I updated to use a public domain image of a cucumber.

swcarter007 commented on packages/allure-cucumberjs/src/CucumberJSAllureReporter.ts:
```diff
@@ -246,7 +246,7 @@ export class CucumberJSAllureFormatter extends Formatter {
 
     if (labels.length > 0) {
-      labels.forEach((label) => this.currentTest?.addLabel(label.name, label.value));
+      labels.forEach((label) => currentTest?.addLabel(label.name, label.value));
```
currentTest wasn't in the right context, which was causing a unit test to fail in the allure-cucumberjs unit test suite.

swcarter007 commented on packages/allure-cucumberjs/src/CucumberJSAllureReporter.ts:
```diff
@@ -334,8 +334,9 @@ export class CucumberJSAllureFormatter extends Formatter {
 
-    const { fileName = "attachment", body, mediaType } = data;
-    const attachmentFilename = this.allureRuntime.writeAttachment(body, mediaType);
+    const { fileName = "attachment", body, mediaType, contentEncoding } = data;
```
CucumberJs passes contentEncoding as either "IDENTITY" or "BASE64" here.  Because I wasn't sure how other libraries like Mocha, Jest, or Playwright responded, I let the default option as 'utf-8'

swcarter007 commented on packages/allure-js-commons/src/writers/AllureWriter.ts:
```diff
@@ -6,7 +6,7 @@ export interface AllureWriter {
   writeGroup(result: TestResultContainer): void;
 
-  writeAttachment(name: string, content: Buffer | string): void;
+  writeAttachment(name: string, content: Buffer | string, encoding?: BufferEncoding): void;
```
Not all usages require the encoding, so it is an optional parameter in the interface.

baev commented on packages/allure-cucumberjs/test/specs/allure_cucumberjs_test.ts:
```diff
@@ -396,4 +417,29 @@ describe("CucumberJSAllureReporter", () => {
   });
+
+
+  it("should add encoding type to images", async () => {
```
as far as I can see test name is wrong

swcarter007 replied:
That whole test was wrong.  I should have deleted it initially as it was a duplicate of the other test.

lamartire commented on packages/allure-js-commons/tsconfig.json:
```diff
@@ -2,7 +2,8 @@
     "./src/**/*",
-    "index.ts"
+    "index.ts",
+    "../allure-cucumberjs/src/BufferEncodingUtility.ts"
```
I'm not sure, that it's good idea to import types from dependent package. Better, to define the type or utility in the `commons` package and then, reuse it in `cucumberjs` one

swcarter007 replied:
This was an accidental include where I placed the file there originally, then I moved it to the other project but it retained this reference.  I have deleted it now.

lamartire commented on packages/allure-cucumberjs/src/BufferEncodingUtility.ts:
```diff
@@ -0,0 +1,8 @@
+export const getBufferEncoding = (encoding: any): BufferEncoding => {
+  switch(encoding) {
+    case "BASE64":
```
Let's use regexp, or turn the argument to lower case to make the function much more generic

swcarter007 replied:
This function was unnecessary as Buffer.isEncoding() already validates a string

lamartire commented on packages/allure-cucumberjs/src/BufferEncodingUtility.ts:
```diff
@@ -0,0 +1,8 @@
+export const getBufferEncoding = (encoding: any): BufferEncoding => {
```
`encoding` should have `string` type

lamartire replied:
And I think, better to rename the function to `normalizeBufferEncoding` because we use it as adapter or normalizer, not getter

swcarter007 replied:
I believe BufferEncoding already is type string.

```
  writeAttachment(name: string, content: Buffer | string, encoding?: BufferEncoding): void {
    const enc = "base64";
    encoding = enc;
    ```
    
    The following works fine.

swcarter007 replied:
unnecessary file

lamartire commented on packages/allure-js-commons/src/AllureRuntime.ts:
```diff
@@ -1,3 +1,4 @@
+import { Encoding } from "crypto";
```
Remove unused code

lamartire commented on packages/allure-js-commons/src/writers/FileSystemAllureWriter.ts:
```diff
@@ -17,9 +17,9 @@ export class FileSystemAllureWriter implements AllureWriter {
   }
 
-  writeAttachment(name: string, content: Buffer | string): void {
+  writeAttachment(name: string, content: Buffer | string, encoding: BufferEncoding = "utf-8"): void {
```
Mark encoding as optional one

lamartire commented on packages/allure-js-commons/src/writers/FileSystemAllureWriter.ts:
```diff
@@ -17,9 +17,9 @@ export class FileSystemAllureWriter implements AllureWriter {
+  writeAttachment(name: string, content: Buffer | string, encoding: BufferEncoding = "utf-8"): void {
     const path = this.buildPath(name);
-    writeFileSync(path, content);
+      writeFileSync(path, content, {encoding: encoding});
```
Just pass `{ encoding }`

swcarter007 commented on packages/allure-cucumberjs/src/CucumberJSAllureReporter.ts:
```diff
@@ -334,8 +333,9 @@ export class CucumberJSAllureFormatter extends Formatter {
-    const { fileName = "attachment", body, mediaType } = data;
-    const attachmentFilename = this.allureRuntime.writeAttachment(body, mediaType);
+    const { fileName = "attachment", body, mediaType, contentEncoding } = data;
+    const encoding = Buffer.isEncoding(contentEncoding) ? contentEncoding : undefined;  // only pass through valid encodings
```
Updated usage of the function to check for a valid encoding.  If valid, it passes it through.  Invalid changes it to undefined for the system to ignore later.

swcarter007 commented:
All fixes have been made and are ready to review again.  Thank you for the short turnaround.  This is blocking our QE department from continuing their testing.

baev reviewed (approved):
LGTM

baev commented:
btw, you can generate a single pixel png image by yourself and use it instead

swcarter007 commented:
> [![CLA assistant check](https://camo.githubusercontent.com/4c970efeed75dd01f6778bf7e33dec85e4171375cb69acb0927bfc9d437d78d9/68747470733a2f2f636c612d617373697374616e742e696f2f70756c6c2f62616467652f6e6f745f7369676e6564)](https://cla-assistant.io/allure-framework/allure-js?pullRequest=468) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/allure-framework/allure-js?pullRequest=468) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/allure-framework/allure-js?pullRequest=468) it.

I'm not sure why this hasn't been processed yet.  I've clicked it a couple of times already and it seemed to work on the website that was opened.

swcarter007 commented:
@lamartire I refactored the code from the original review and deleted the code that you had requested changes on.  When you have a moment, can you take a look again to see if everything looks okay?  I appreciate the reviews.

lamartire commented:
> @lamartire I refactored the code from the original review and deleted the code that you had requested changes on. When you have a moment, can you take a look again to see if everything looks okay? I appreciate the reviews.

Looking right now 👍

lamartire commented on packages/allure-js-commons/src/writers/FileSystemAllureWriter.ts:
```diff
@@ -17,9 +17,13 @@ export class FileSystemAllureWriter implements AllureWriter {
+  writeAttachment(name: string, content: Buffer | string, encoding?: BufferEncoding): void {
     const path = this.buildPath(name);
-    writeFileSync(path, content);
+    if (encoding) {
```
I think, the condition is not needed here, let's set `utf-8` as default encoding and make the logic much more linear

swcarter007 replied:
Updated.

lamartire reviewed (changes requested):
Most of the code looks nice, there left some little things to fix. 
Also, I can't compile the library:

```shell
λ yarn compile
src/CucumberJSAllureReporter.ts:338:84 - error TS2554: Expected 2 arguments, but got 3.

338     const attachmentFilename = this.allureRuntime.writeAttachment(body, mediaType, encoding);
                                                                                       ~~~~~~~~


Found 1 error.
```

Please, take a look and fix it

swcarter007 commented:
> 

I ran `yarn compile` on the root directory, the allure-cucumberjs, and the allure-js-commons directory locally and didn't see any error here.

lamartire reviewed (approved):
👍 good job, no more issues here from my side

lamartire commented:
> > 
> 
> I ran `yarn compile` on the root directory, the allure-cucumberjs, and the allure-js-commons directory locally and didn't see any error here.

My fault, `commons` package must be compiled first

lamartire commented:
Please, check linters issues

swcarter007 commented:
> Please, check linters issues

I get this when I run it on the root
![image](https://user-images.githubusercontent.com/16365812/181292073-83c82a46-9022-4fa2-af46-1c1af8488b92.png)

swcarter007 commented:
> Please, check linters issues

![image](https://user-images.githubusercontent.com/16365812/181292498-67df4fe0-24a8-4bcd-84e7-43f1398d338a.png)

lamartire commented:
> > Please, check linters issues
> 
> ![image](https://user-images.githubusercontent.com/16365812/181292498-67df4fe0-24a8-4bcd-84e7-43f1398d338a.png)

https://github.com/allure-framework/allure-js/runs/7543818268?check_suite_focus=true

swcarter007 commented:
> 

It looks to be a prettier formatting issue, but when I run that locally against the root I receive that no files were found.

![image](https://user-images.githubusercontent.com/16365812/181298858-533a5657-cd4b-4955-8e83-d83500547044.png)

swcarter007 commented:
This PR is ready to merge by an authrozied reviewer.  Thanks for the reviews and feedback.

baev commented:
@swcarter007 thanks!
